app/scripts/services/oauth/oauth-service.ts

```diff
@@ -0,0 +1,215 @@
+   * @returns The new refresh token.
+   */
+  async getNewRefreshToken(options: {
+    connection: AuthConnection;
+    refreshToken: string;
+  }): Promise<{ idTokens: string[] }> {
+    const { connection, refreshToken } = options;
+    const loginHandler = createLoginHandler(
+      connection,
+      this.#env,
+      this.#webAuthenticator,
+    );
+
+    const refreshTokenData = await loginHandler.refreshAuthToken(refreshToken);
+    const idToken = refreshTokenData.jwt_tokens[this.#audience];
+
+    return {
+      idTokens: [idToken],
+    };
+  }
+
+  /**
+   * Revoke the current refresh token and get a new refresh token.
+   *
+   * @param options - The options for the revoke and get new refresh token.
+   * @param options.connection - The social login type to login with.
+   * @param options.revokeToken - The revoke token to authenticate the revoke request.
+   * @returns The new refresh token and revoke token.
+   */
+  async revokeAndGetNewRefreshToken(options: {
+    connection: AuthConnection;
+    revokeToken: string;
+  }): Promise<{ newRevokeToken: string; newRefreshToken: string }> {
+    const { connection, revokeToken } = options;
+    const loginHandler = createLoginHandler(
+      connection,
+      this.#env,
+      this.#webAuthenticator,
+    );
+
+    const res = await loginHandler.revokeRefreshToken(revokeToken);
+    return {
+      newRefreshToken: res.refresh_token,
+      newRevokeToken: res.revoke_token,
+    };
+  }
+
+  /**
+   * Handle the OAuth login for the given social login type.
+   *
+   * For Google login, we will use the `PKCE` flow to get the Authorization Code.
+   * For Apple login, we will use the `server-redirect` flow to get the Authorization Code.
+   *
+   * Then, we will use the Authorization Code to get the Jwt Token from the Web3Auth Authentication Server.
+   *
+   * @param loginHandler - The login handler to use.
+   * @returns The login result.
+   */
+  async #handleOAuthLogin(loginHandler: BaseLoginHandler) {
+    const authUrl = await loginHandler.getAuthUrl();
+
+    // launch the web auth flow to get the Authorization Code from the social login provider
+    const redirectUrlFromOAuth = await new Promise<string>(
+      (resolve, reject) => {
+        // since promise returns aren't supported until MV3, we need to use a callback function to support MV2
+        this.#webAuthenticator.launchWebAuthFlow(
+          {
+            interactive: true,
+            url: authUrl,
+          },
+          (responseUrl) => {
+            try {
+              if (responseUrl) {
+                const url = new URL(responseUrl);
+                const state = url.searchParams.get('state');
+
+                loginHandler.validateState(state);
+                resolve(responseUrl);
+              } else {
+                reject(new Error('No redirect URL found'));
+              }
+            } catch (error: unknown) {
+              reject(error);
```

Comment:
handle state validation error here and throw a proper error msg.. probably from an error class